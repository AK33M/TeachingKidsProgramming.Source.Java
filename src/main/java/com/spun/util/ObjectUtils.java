package com.spun.util;

import org.teachingkidsprogramming.util.ArrayUtils;
import org.teachingkidsprogramming.util.ClassUtils;

import java.lang.reflect.Array;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * A static class of convenience functions for Manipulating objects
 **/
public class ObjectUtils {
	public static Method[] getMethodsForObject(Object o2, String[] passedMethods)
			throws SecurityException, NoSuchMethodException {
		Method methods[] = new Method[passedMethods.length];
		Class clazz = o2.getClass();
		for (int i = 0; i < passedMethods.length; i++) {
			methods[i] = clazz.getMethod(passedMethods[i], (Class[]) null);
		}
		return methods;
	}

	/**
	 * A convenience function to check if two objects are equal.
	 * 
	 * @param s1
	 *            The first object
	 * @param s2
	 *            The second object
	 * @return true if Equal.
	 **/
	public static boolean isEqual(Object s1, Object s2) {
		if (s1 == s2) {
			return true;
		} else if ((s1 != null) && s1.equals(s2)) {
			return true;
		} else {
			return false;
		}
	}

	public static boolean isIn(Object target, Object[] objects) {
		for (int i = 0; i < objects.length; i++) {
			if (ObjectUtils.isEqual(objects[i], target)) {
				return true;
			}
		}
		return false;
	}

	public static boolean isThisInstanceOfThat(Class<?> thiz, Class<?> that) {
		return that.isAssignableFrom(thiz);
	}

	public static Error throwAsError(Throwable t) throws Error {
		if (t instanceof RuntimeException) {
			throw (RuntimeException) t;
		} else if (t instanceof Error) {
			throw (Error) t;
		} else {
			throw new Error(t);
		}
	}

	/**
	 * @deprecated use Query.select()
	 */
	public static Object[] extractArray(Object[] from, String methodName) {
		try {
			if (from == null || from.length == 0) {
				return new Object[0];
			}
			Method method = getGreatestCommonDenominator(from, methodName);
			Object[] array = null;
			if (Object.class.isAssignableFrom(method.getReturnType())) {
				array = (Object[]) Array.newInstance(method.getReturnType(),
						from.length);
			} else {
				array = (Object[]) Array.newInstance(
						ClassUtils.getWrapperClass(method.getReturnType()),
						from.length);
			}
			for (int i = 0; i < from.length; i++) {
				array[i] = method.invoke(from[i], (Object[]) null);
			}
			return array;
		} catch (Exception e) {
			MySystem.warning(e);
			throw ObjectUtils.throwAsError(e);
		}
	}

	public static Method getGreatestCommonDenominator(Object[] from,
			String methodName) throws SecurityException, NoSuchMethodException {
		List<Class> classes = new ArrayList<Class>();
		ArrayUtils.addArray(classes, getAllCastableClasses(from[0]));
		for (Object o : from) {
			for (int i = classes.size() - 1; i >= 0; i--) {
				Class clazz = classes.get(i);
				if (!isThisInstanceOfThat(o.getClass(), clazz)
						|| !ClassUtils.hasMethod(clazz, methodName)) {
					classes.remove(i);
				}
			}
		}
		return classes.size() == 0 ? null : ArrayUtils.getLast(classes)
				.getMethod(methodName, (Class[]) null);
	}

	private static Class[] getAllCastableClasses(Object object) {
		Class<? extends Object> clazz = object.getClass();
		ArrayList<Object> list = new ArrayList<Object>();
		while (clazz != null) {
			list.add(clazz);
			ArrayUtils.addArray(list, clazz.getInterfaces());
			clazz = clazz.getSuperclass();
		}
		Class[] found = (Class[]) list.toArray(new Class[list.size()]);
		ArrayUtils.toReverseArray(found);
		return found;
	}

	public static void assertInstance(Class clazz, Object object) {
		assertInstance(new Class[] { clazz }, object);
	}

	public static void assertInstance(Class classes[], Object object) {
		if (object == null) {
			throw new NullPointerException("Expected Object of Type "
					+ Arrays.asList(extractArray(classes, "getName"))
					+ " but was null");
		}
		for (int i = 0; i < classes.length; i++) {
			if (ClassUtils.getWrapperClass(classes[i]).isInstance(object)) {
				return;
			}
		}
		throw new IllegalArgumentException("Expected Object of Type "
				+ Arrays.asList(extractArray(classes, "getName")) + " but got "
				+ object.getClass().getName());
	}

	public static String getClassName(Object o) {
		return o == null ? "null" : o.getClass().getName();
	}
}