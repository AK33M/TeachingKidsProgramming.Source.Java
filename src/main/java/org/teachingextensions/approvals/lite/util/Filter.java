package org.teachingextensions.approvals.lite.util;

public interface Filter<T> {
    /**
     * @param object the item to check
     * @return true if the object would be extracted by the filter
     * @throws IllegalArgumentException if the object is not supported by the filter
     */
    public boolean isExtracted(T object)
            throws IllegalArgumentException;
}