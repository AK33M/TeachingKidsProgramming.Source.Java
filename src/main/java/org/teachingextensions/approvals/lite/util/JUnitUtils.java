package org.teachingextensions.approvals.lite.util;

import java.awt.GraphicsEnvironment;
import static org.junit.Assume.*;

public class JUnitUtils {
    public static void assumeNotHeadless() {
        boolean headless_check = GraphicsEnvironment.isHeadless();
        assumeFalse(headless_check);
    }
}
