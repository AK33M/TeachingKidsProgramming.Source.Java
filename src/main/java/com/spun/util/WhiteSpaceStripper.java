package com.spun.util;

import org.teachingkidsprogramming.util.MySystem;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

public class WhiteSpaceStripper 
{
	/***********************************************************************/
	public static void stripFolder(File dir)
  {
    stripFolder(dir, true);
  }
	/***********************************************************************/
	public static void stripFolder(File dir, boolean recursive)
  {
  	if (!dir.isDirectory())
    {
    	MySystem.warning("File is not a Directory - " + dir.toString());
    	return;
    }
  	
  	File[] files = dir.listFiles(new WhiteSpaceFileFilter());
  	for (int i = 0; i < files.length; i++)
    {
    	if (files[i].isDirectory())
      {
      	MySystem.event("Scaning Directory -" + files[i].getName());
      	if (recursive) stripFolder(files[i], recursive);
      }
    	else
      {
        stripFile(files[i]);
      }
    }
  }

    /***********************************************************************/
	public static void stripFile(File file)
  {
    
  	if (!file.isFile())
    {
    	MySystem.warning("File is not a File - " + file.toString());
    	return;
    }
		if(!file.canWrite())
		{
			MySystem.event("File '" + file.toString() + "' is readonly");
        }
		else
		{
	    try
	    {
	      String contents = readFile(file);
	      String stripped = stripWhiteSpace(contents);
	      writeFile(file, stripped);
	    }
	    catch (IOException e)
	    {
	      MySystem.warning(e);
	    }
		}
  }
  /***********************************************************************/
  public static String stripWhiteSpace(String text)
  {
    StringBuilder newText = new StringBuilder();
    boolean whitespace = false;
    int num = text.length();
		char whiteSpaceChar = ' ';
    for (int i = 0; i < num; i++)
    {
      char c = text.charAt(i);
			switch (c)
			{
				case '\n' : whiteSpaceChar = '\n';whitespace = true;break;
				case '\t' :
				case ' '  : whitespace = true;break;
				default	  : 
					if (whitespace)
					{
						whitespace = false;
						newText.append(whiteSpaceChar);
						whiteSpaceChar = ' '; 
					}						
	        newText.append(c);break;
      }
    }
    return newText.toString();
  }

    /***********************************************************************/
  private static String readFile(File file)
  throws IOException
  {
    BufferedReader reader = new BufferedReader(new FileReader(file));
    StringBuilder output = new StringBuilder();
    
    while (reader.ready())
    {
      output.append(reader.readLine());
	    output.append("\n");
    }
    
    reader.close();
    return output.toString();
  }
	/***********************************************************************/
	private static void writeFile(File file, String text)
	throws IOException
	{
    FileWriter writer = new FileWriter(file);
    
    writer.write(text);
    
    writer.close();
	}
	/***********************************************************************/
  public static void main (String [] args)
  {
//    stripFolder(new File("C:\\temp\\stockgazing"));
  }
	/***********************************************************************/
  /***********************************************************************/
}


class WhiteSpaceFileFilter
  implements java.io.FileFilter
{

  /***********************************************************************/

  public boolean accept(File pathname)
  {
  	
    if (pathname.getName().equals(".") || pathname.getName().equals("."))
    {
      return false;
    }
    else if (pathname.getName().equalsIgnoreCase("email"))
    {
      return false;
    }
    else
        return pathname.isDirectory() || (pathname.getName().indexOf(".htm") != -1) || (pathname.getName().indexOf(".txt") != -1);
  }
  
  /***********************************************************************/
  /***********************************************************************/
}