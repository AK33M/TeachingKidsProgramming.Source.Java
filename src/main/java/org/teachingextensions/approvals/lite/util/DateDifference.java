package org.teachingextensions.approvals.lite.util;

import java.util.Calendar;
import java.util.Date;

public class DateDifference {
    public static final String STANDARD_TIME_TEXT[] = {"Year", "Years", "Month", "Months", "Week", "Weeks", "Day", "Days", "Hour", "Hours", "Min", "Mins", "Sec", "Secs", "Milli", "Millis"};
    public static final String MILLISECONDS = "milliseconds";
    public static final String SECONDS = "seconds";
    public static final String MINUTES = "minutes";
    public static final String HOURS = "hours";
    public static final String DAYS = "days";
    public static final String WEEKS = "weeks";
    public static final String MONTHS = "months";
    public static final String YEARS = "years";
    public static String TIME_UNITS[] = {YEARS, MONTHS, WEEKS, DAYS, HOURS, MINUTES, SECONDS, MILLISECONDS};
    // assumes a 30 day month
    private static int TIME_SCALE[] = {Calendar.YEAR, Calendar.MONTH, Calendar.WEEK_OF_YEAR, Calendar.DATE, Calendar.HOUR, Calendar.MINUTE, Calendar.SECOND, Calendar.MILLISECOND};
    private static long DIVIDERS[] = {1000 * 60 * 60 * 24 * 365L, 1000 * 60 * 60 * 24 * 30L, 1000 * 60 * 60 * 24 * 7L, 1000 * 60 * 60 * 24L, 1000 * 60 * 60L, 1000 * 60L, 1000L, 1L};
    /**
     * ********************************************************************
     */
    private long milli;

    /**
     * ********************************************************************
     */
    public DateDifference(long timeDifference) {
        this.milli = timeDifference;
    }

    /**
     * ********************************************************************
     */
    public DateDifference(Date date1, Date date2) {
        milli = date1.getTime() - date2.getTime();
        if (milli < 0) {
            milli = milli * -1;
        }
    }

    /**
     * ********************************************************************
     */
    public static long getAbsoluteDifference(int unit, long time) {
        return time / DIVIDERS[getTimeScaleIndex(unit)];
    }

    /**
     * ********************************************************************
     */
    public static long getRemainingDifference(int wantedUnit, int roundTo, long time) {
        int wantedIndex = getTimeScaleIndex(wantedUnit);
        int roundToIndex = getTimeScaleIndex(roundTo);
        if (wantedIndex < roundToIndex) {
            throw new Error("Can't round with a smaller Unit.[" + TIME_UNITS[wantedIndex] + ", " + TIME_UNITS[roundToIndex] + "]");
        }
        return (time % DIVIDERS[roundToIndex]) / DIVIDERS[wantedIndex];
    }
    /***********************************************************************/
    /**
     * @return the index in TIME_SCALE[] of the largest Unit to be > 0
     */
    private int getMaximumTimeUnit() {
        int i = 0;
        while ((i < DIVIDERS.length) && (milli < DIVIDERS[i])) {
            i++;
        }
        return i;
    }
    /***********************************************************************/
    /**
     * Finds the index for a Calendar.DATE ect.
     *
     * @return the index in TIME_SCALE[]
     */
    public static int getTimeScaleIndex(int calendarTime) {
        int i = 0;
        while (calendarTime != TIME_SCALE[i]) {
            i++;
        }
        return i;
    }

    /**
     * *******************************************************************
     */
    private static long getStandardRoundedTime(int unitIndex, boolean forceAbsolute, long time) {
        if (unitIndex == 0 || forceAbsolute) {
            return getAbsoluteDifference(TIME_SCALE[unitIndex], time);
        } else {
            return getRemainingDifference(TIME_SCALE[unitIndex], TIME_SCALE[unitIndex - 1], time);
        }
    }

    /**
     * *******************************************************************
     */
    public String getTimeText(int amount, int maxUnit, int minUnit, String nowText, String agoText, String units[]) {
        if (amount == 0) {
            throw new Error("getTimeText() requires amount > 0");
        }
        maxUnit = getTimeScaleIndex(maxUnit);
        minUnit = getTimeScaleIndex(minUnit);
        int realMax = getMaximumTimeUnit();
        String timeText = nowText;
        if (realMax < maxUnit) {
            realMax = maxUnit;
        }
        if (realMax <= minUnit) {
            timeText = "";
            long remainingTime = this.milli;
            for (int i = realMax; i < (realMax + amount) && (i <= minUnit); i++) {
                long time = getStandardRoundedTime(i, i == realMax, remainingTime);
                remainingTime -= time * DIVIDERS[i];
                timeText += time + " " + units[(time == 1) ? i * 2 : i * 2 + 1] + ", ";
            }
            timeText = timeText.substring(0, timeText.length() - 2);
            if ((agoText != null) && (agoText.length() > 0)) {
                timeText += " " + agoText;
            }
        }
        return timeText;
    }

    /**
     * *******************************************************************
     */
    public String getStandardTimeText(int amountShown) {
        return getTimeText(amountShown, Calendar.YEAR, Calendar.MILLISECOND, "now", "", STANDARD_TIME_TEXT);
    }
    /************************************************************************/
    /************************************************************************/
}