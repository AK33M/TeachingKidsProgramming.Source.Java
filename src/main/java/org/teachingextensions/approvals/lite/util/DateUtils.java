package org.teachingextensions.approvals.lite.util;

import java.sql.Timestamp;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

/**
 * A static class of convenience functions for database access
 */
public class DateUtils {
    private static int TIME_SCALE[] = {Calendar.YEAR,
            Calendar.MONTH,
            Calendar.DATE,
            Calendar.HOUR_OF_DAY,
            Calendar.HOUR,
            Calendar.MINUTE,
            Calendar.SECOND,
            Calendar.MILLISECOND};

    public static boolean isSame(Date firstDate, Date secondDate, int smallestUnits) {
        if ((firstDate == null) || (secondDate == null)) {
            return (firstDate == secondDate);
        }
        return isSame(firstDate.getTime(), secondDate.getTime(), smallestUnits);
    }

    public static boolean isSame(long firstDate, long secondDate, int smallestUnits) {
        if (!ArrayUtils.contains(TIME_SCALE, smallestUnits)) {
            throw new Error("Invalid Timescale " + smallestUnits);
        }
        GregorianCalendar first = new GregorianCalendar();
        first.setTime(new Date(firstDate));
        setSignificantDigit(first, smallestUnits);
        GregorianCalendar second = new GregorianCalendar();
        second.setTime(new Date(secondDate));
        setSignificantDigit(second, smallestUnits);
        return (first.getTime().getTime() == second.getTime().getTime());
    }

    /**
     * ********************************************************************
     */
    public static Timestamp getStartOfXDaysAgo(int numberOfDays) {
        return getStartOfXDaysAgo(numberOfDays, new Date());
    }

    /**
     * ********************************************************************
     */
    public static Timestamp getStartOfXDaysAgo(int numberOfDays, Date startingFrom) {
        GregorianCalendar gregorianCalendar = new GregorianCalendar();
        gregorianCalendar.setTime(startingFrom);
        DateUtils.setSignificantDigit(gregorianCalendar, Calendar.DATE);
        gregorianCalendar.add(Calendar.DATE, -numberOfDays);
        return new Timestamp(gregorianCalendar.getTime().getTime());
    }

    /**
     * ********************************************************************
     */
    public static Calendar setSignificantDigit(Calendar calendar, int smallestUnits) {
        boolean removeOn = false;
        for (int i : TIME_SCALE) {
            if (removeOn) {
                if (i == Calendar.DATE) {
                    // Work around since this is broke
                    calendar.set(i, 1);
                } else if (i == Calendar.HOUR_OF_DAY) {
                    // Work around since this is broke
                    calendar.set(i, 0);
                } else {
                    calendar.clear(i);
                }
            }
            if (i == smallestUnits) {
                removeOn = true;
            }
        }
        return calendar;
    }

    public static void main(String args[]) {
        MySystem.variable("Calendar.DATE = " + Calendar.DATE);
        MySystem.variable("Year", setSignificantDigit(new GregorianCalendar(), Calendar.YEAR).getTime());
        MySystem.variable("Month", setSignificantDigit(new GregorianCalendar(), Calendar.MONTH).getTime());
        MySystem.variable("Day", setSignificantDigit(new GregorianCalendar(), Calendar.DAY_OF_MONTH).getTime());
        MySystem.variable("Hour", setSignificantDigit(new GregorianCalendar(), Calendar.HOUR).getTime());
        MySystem.variable("Minute", setSignificantDigit(new GregorianCalendar(), Calendar.MINUTE).getTime());
        MySystem.variable("End Of Day", rollToEndOfDay(new Date()).getTime());
    }

    /**
     * ********************************************************************
     */
    public static GregorianCalendar rollToEndOfDay(Date date) {
        GregorianCalendar gregorianCalendar = new GregorianCalendar();
        gregorianCalendar.setTime(date);
        gregorianCalendar.set(Calendar.HOUR_OF_DAY, 23);
        gregorianCalendar.set(Calendar.MINUTE, 59);
        gregorianCalendar.set(Calendar.SECOND, 59);
        gregorianCalendar.set(Calendar.MILLISECOND, 999);
        return gregorianCalendar;
    }

    /**
     * *******************************************************************
     */
    public static Timestamp asTimestamp(Date date) {
        return new Timestamp(date.getTime());
    }

    /**
     * ********************************************************************
     */
    public static boolean isToday(Date date) {
        return DateUtils.isSame(date, new Date(), Calendar.DATE);
    }

    /**
     * ********************************************************************
     */
    public static Calendar asCalendar(Date date) {
        GregorianCalendar gregorianCalendar = new GregorianCalendar();
        gregorianCalendar.setTime(date);
        return gregorianCalendar;
    }

    /************************************************************************/
    /**
     * @param date "yyyy/MM/dd"
     */
    public static Timestamp parse(String date) {
        SimpleDateFormat format = new java.text.SimpleDateFormat("yyyy/MM/dd");
        try {
            return asTimestamp(format.parse(date));
        } catch (ParseException e) {
            throw ObjectUtils.throwAsError(e);
        }
    }
    /************************************************************************/
}