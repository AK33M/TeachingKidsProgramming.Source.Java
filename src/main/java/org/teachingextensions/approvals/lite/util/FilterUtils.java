package org.teachingextensions.approvals.lite.util;

import java.util.ArrayList;
import java.util.Arrays;

/**
 * Listens to the state of a EnabledConditions object
 */
public class FilterUtils {
    /**
     * @param fromObjects a collection of objects to filter
     * @param filter an optional parameter which indicates whether the item is extracted
     * @return a new List containing all elements of the list for which isExtracted() would return true
     */
    public static <T> ArrayList<T> retainExtracted(T fromObjects[], Filter filter) throws IllegalArgumentException {
        return fromObjects == null ? new ArrayList<T>() : filter(Arrays.asList(fromObjects), filter, true);
    }

    /**
     * @param collection      the collection to filter
     * @param filter          an optional parameter which indicates whether the item is extracted
     * @param retainExtracted true if the filter should retained the extracted items
     * @return a new List containing all elements of the list for which isExtracted() would return true
     */
    private static <T> ArrayList<T> filter(Iterable<? extends T> collection, Filter filter, boolean retainExtracted) throws IllegalArgumentException {
        ArrayList<T> extracted = new ArrayList<>();
        if (collection != null) {
            for (T t : collection) {
                boolean ex = (filter == null || filter.isExtracted(t));
                if (ex == retainExtracted) {
                    extracted.add(t);
                }
            }
        }
        return extracted;
    }
}