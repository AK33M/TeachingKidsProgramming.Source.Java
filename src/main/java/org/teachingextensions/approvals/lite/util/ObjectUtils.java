package org.teachingextensions.approvals.lite.util;

import java.lang.reflect.Array;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * A static class of convenience functions for Manipulating objects
 */
public class ObjectUtils {

    /**
     * A convenience function to check if two objects are equal.
     *
     * @param s1 The first object
     * @param s2 The second object
     * @return true if Equal.
     */
    public static boolean isEqual(Object s1, Object s2) {
        return s1 == s2 || (s1 != null) && s1.equals(s2);
    }

    public static boolean isThisInstanceOfThat(Class<?> thiz, Class<?> that) {
        return that.isAssignableFrom(thiz);
    }

    public static Error throwAsError(Throwable t) throws Error {
        if (t instanceof RuntimeException) {
            throw (RuntimeException) t;
        } else if (t instanceof Error) {
            throw (Error) t;
        } else {
            throw new Error(t);
        }
    }

    /**
     * @param from       the source array
     * @param methodName the filter method
     * @return a filtered array
     */
    public static Object[] extractArray(Object[] from, String methodName) {
        try {
            if (from == null || from.length == 0) {
                return new Object[0];
            }
            Method method = getGreatestCommonDenominator(from, methodName);
            Object[] array;
            if (Object.class.isAssignableFrom(method.getReturnType())) {
                array = (Object[]) Array.newInstance(method.getReturnType(),
                        from.length);
            } else {
                array = (Object[]) Array.newInstance(
                        ClassUtils.getWrapperClass(method.getReturnType()),
                        from.length);
            }
            for (int i = 0; i < from.length; i++) {
                array[i] = method.invoke(from[i], (Object[]) null);
            }
            return array;
        } catch (Exception e) {
            MySystem.warning(e);
            throw ObjectUtils.throwAsError(e);
        }
    }

    public static Method getGreatestCommonDenominator(Object[] from,
                                                      String methodName) throws SecurityException, NoSuchMethodException {
        List<Class> classes = new ArrayList<>();
        ArrayUtils.addArray(classes, getAllCastableClasses(from[0]));
        for (Object o : from) {
            for (int i = classes.size() - 1; i >= 0; i--) {
                Class clazz = classes.get(i);
                if (!isThisInstanceOfThat(o.getClass(), clazz)
                        || !ClassUtils.hasMethod(clazz, methodName)) {
                    classes.remove(i);
                }
            }
        }
        return classes.size() == 0 ? null : ArrayUtils.getLast(classes)
                .getMethod(methodName, (Class[]) null);
    }

    private static Class[] getAllCastableClasses(Object object) {
        Class<?> clazz = object.getClass();
        ArrayList<Object> list = new ArrayList<>();
        while (clazz != null) {
            list.add(clazz);
            ArrayUtils.addArray(list, clazz.getInterfaces());
            clazz = clazz.getSuperclass();
        }
        Class[] found = list.toArray(new Class[list.size()]);
        ArrayUtils.toReverseArray(found);
        return found;
    }

    public static void assertInstance(Class clazz, Object object) {
        assertInstance(new Class[]{clazz}, object);
    }

    public static void assertInstance(Class classes[], Object object) {
        if (object == null) {
            throw new NullPointerException("Expected Object of Type "
                    + Arrays.asList(extractArray(classes, "getName"))
                    + " but was null");
        }
        for (Class aClass : classes) {
            if (ClassUtils.getWrapperClass(aClass).isInstance(object)) {
                return;
            }
        }
        throw new IllegalArgumentException("Expected Object of Type "
                + Arrays.asList(extractArray(classes, "getName")) + " but got "
                + object.getClass().getName());
    }

    public static String getClassName(Object o) {
        return o == null ? "null" : o.getClass().getName();
    }
}