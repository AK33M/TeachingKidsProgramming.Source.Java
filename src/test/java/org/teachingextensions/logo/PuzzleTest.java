package org.teachingextensions.logo;

import org.junit.Test;

import java.util.Arrays;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class PuzzleTest {
  private static int[] getSolution() {
    return new int[]{0, 1, 2, 3, 4, 5, 6, 7, 8};
  }

  private static int[] swap(int[] cells, int i, int j) {
    int[] copy = Arrays.copyOf(cells, cells.length);
    int x = copy[i];
    copy[i] = copy[j];
    copy[j] = x;
    return copy;
  }

  /**
   * A puzzle has a solution
   */
  @Test
  public void puzzle_has_solution() {
    int[] solved = getSolution();
    Puzzle p = new Puzzle(solved);
    assertTrue(p.isSolved());
  }

  /**
   * A puzzle might not be solved
   */
  @Test
  public void puzzle_not_solved() {
    int[] cells = getSolution();
    cells = swap(cells, 3, 4);
    Puzzle p = new Puzzle(cells);
    assertFalse(p.isSolved());
  }

  /**
   * A puzzle has a blank
   */
  @Test
  public void puzzle_has_blank(){
    Puzzle p = new Puzzle(getSolution());
    assertEquals(8, p.getBlankIndex());
  }

  /**
   * Blank can be anywhere
   */
  @Test
  public void blank_can_be_anywhere(){
    int[] cells = swap(getSolution(), 8, 2);
    Puzzle p = new Puzzle(cells);
    assertEquals(2, p.getBlankIndex());
  }

  /**
   * Puzzle can swap blank for you
   */
  @Test
  public void puzzle_can_swap_blank(){
    Puzzle p = new Puzzle(getSolution());
    Puzzle c = p.swapBlank(4);
    assertEquals(4, c.getBlankIndex());
  }

  /**
   * Puzzle can swap blank for you
   */
  @Test
  public void blank_swap_leaves_original_as_is(){
    Puzzle p = new Puzzle(getSolution());
    p.swapBlank(4);
    assertEquals(8, p.getBlankIndex());
  }

}
