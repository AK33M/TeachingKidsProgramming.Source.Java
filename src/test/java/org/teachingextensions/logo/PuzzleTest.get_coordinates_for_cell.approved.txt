Positions


java.awt.Point[x=0,y=0]
java.awt.Point[x=1,y=1]
java.awt.Point[x=2,y=2]
java.awt.Point[x=1,y=0]
